orangecontrib/imageanalytics/widgets/owimageviewer.py

```diff
@@ -311,38 +306,59 @@ def sizeHint(self):
 
     @Inputs.data
     def setData(self, data):
-        self.closeContext()
+        if self.image_attr is not None:
+            # Don't store invalid contexts because they will match anything
+            # and crash the widget when they're used.
+            self.closeContext()
         self.clear()
-        self.data = data
-
-        if data is not None:
-            self.image_model.set_domain(data.domain)
-            self.title_model.set_domain(data.domain)
-            im_attr = filter_image_attributes(self.data)
-            self.image_attr = im_attr[0] if im_attr else None
-            self.title_attr = self.title_model[0] if self.title_model else None
+        if data is None:
+            self.commit.now()
+            return
 
-            self.openContext(data)
+        self.image_model.set_domain(data.domain)
+        self.title_model.set_domain(data.domain)
+        if not self.image_model:
+            self.Error.no_image_attr()
+            self.commit.now()
+            return
 
-            if self.image_model:
-                self.setupModel()
+        self.data = data
+        self._propose_image_and_title_attr()
+        self.openContext(data)
+        self.setupModel()
         self.commit.now()
 
-    def __select_image_attr(self):
-        for attr in self.image_model:
-            if attr.attributes.get("type").lower() == "image":
-                return attr
-        # check if function already exist
-        return self.image_model[0] if self.image_model else None
+    def _propose_image_and_title_attr(self):
+        self.image_attr = max(
+            self.image_model,
+            key=lambda attr: attr.attributes.get("type", "").lower() == "image"
+        )
+        # Use class variable if it exists. Otherwise,
+        # prefer string variables (there will be at least one, otherwise
+        # image_model is empty and widget reports an error,
+        # but avoid those marked as "image" and in particular the one used
+        # for image_attr
+        self.title_attr = self.data.domain.class_var or max(
```

Comment:
Why class var with the highest priority? Is it really more relevant than the image title?

Reply:
I'm just keeping the existing behavior. Also because of all existing guides and possibly videos.

It seemed weird to me, too, but it makes sense in one very common setup: if you load images from subdirectories, subdirectory names become classes, and you may want those see class names (e.g. types of handbags, or names of trees, or types of buildings...) as titles.